### PR TITLE
core: stdcm: use standard allowance in heuristic

### LINKED ISSUE #13568: core: stdcm: use standard allowance in heuristic
It would make the heuristic more accurate and improve performances

### PR DESCRIPTION
**Pull request by eckter**
Fix https://github.com/OpenRailAssociation/osrd/issues/13568

The heuristic was always optimistic by at least the allowance value. This should bring it closer to the actual shortest time. It only makes a difference when there's an allowance, but that should always be the case in prod. 

Also included in a second commit: we now prioritize nodes that are further from the origin when everything else is equal. This avoids exploring several parallel paths at the same time.

My benchmark is weaker than before as I "lost" the payload collection from before the timetable cache refactoring. But on a smaller sample of 8 requests, average computation time went from 14 to 10s. 

Notes:

* We can also consider adding braking curves to the heuristic envelopes. It would get *really close* to the actual fastest possible path, but computing the heuristic would also be slower. 
* I've encountered some memory issues when looking for long requests. Some requests took *a lot* longer when core has the same memory limit as in prod (5gb), I've also had one actual OOM. We could investigate further.
* I've seen an "f-value increasing" error **once** and did not manage to reproduce it... In doubt, a third commit aggregates these errors, to avoid cluttering the logs if it happens in prod.

### CODE REVIEW

File: core/src/main/kotlin/fr/sncf/osrd/utils/CachedBlockMRSPBuilder.kt [hunk 1]
```diff
@@ -46,6 +47,17 @@ data class CachedBlockMRSPBuilder(
 
     /** Returns the speed limits for the given block (cached). */
     fun getMRSP(block: BlockId): Envelope {
+        if (block.index == 33857U) {
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/utils/CachedBlockMRSPBuilder.kt]: Looks like a debug leftover.

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMNode.kt [hunk 2]
```diff
@@ -83,12 +83,19 @@ data class STDCMNode(
...
+        // Prioritize the highest number of reached targets, to avoid paths that have missed a step
+        else if (
+            other.infraExplorer.getStepTracker().nStepsExcludingLookahead !=
+                infraExplorer.getStepTracker().nStepsExcludingLookahead
+        )
             other.infraExplorer.getStepTracker().nStepsExcludingLookahead -
                 infraExplorer.getStepTracker().nStepsExcludingLookahead
+
+        // Prioritize the path furthest from the start. Avoids exploring several paths in parallel.
+        else -timeData.timeSinceDeparture.compareTo(other.timeData.timeSinceDeparture)
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMNode.kt:98]: If I understand correctly: prioritize path where we were able to be as good as the prediction-heuristic (which is supposed optimistic).
Makes sense 👍

File: core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/STDCMHeuristicTests.kt [hunk 3]
```diff
@@ -80,6 +80,7 @@ class STDCMHeuristicTests {
                     Double.POSITIVE_INFINITY,
                     STANDARD_TRAIN,
                     mrspBuilder = CachedBlockMRSPBuilder(infra, infra, null),
+                    allowance = null,
```
Comment (bougue-pe) [inline core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/STDCMHeuristicTests.kt:84]: Can it be tested with an actual allowance? (mostly to insure no crash in the future, so maybe testing corner cases also could be cheap and good)

Maybe there are already existing cases for the whole STDCM tests, though.

**Review by bougue-pe — APPROVED** [review]
It looks good.

File: core/src/main/kotlin/fr/sncf/osrd/utils/CachedBlockMRSPBuilder.kt (hunk 1 above)
Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/utils/CachedBlockMRSPBuilder.kt]: It definitely is 😅 I pushed all my code at some point to transfert it to my laptop, so I didn't proofread it as part of my first `commit -v` like usual

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMNode.kt (hunk 2 above)
Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMNode.kt:98]: It's more specific than that, it's a small optimization mentioned on [the wikipedia page of A*](https://en.wikipedia.org/wiki/A*_search_algorithm#Implementation_details). When there are several solutions of equal costs, depending on how we break the ties, we may end up exploring them all in a "breadth-first" way. The version here would result in more of a "depth-first" traversal, where we could avoid visiting redundant paths. 

I benched this change on its own but in a more superficial way, it seems to improves performances but only by a few percents. Not much but good enough for such a small change.

File: core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/STDCMHeuristicTests.kt [hunk 4]
```diff
@@ -109,6 +111,64 @@ class STDCMHeuristicTests {
         assertEquals(0.0, getLocationRemainingTime(infra, explorer, null, heuristic))
     }
 
+    @Test
+    fun allowanceTest() {
```
Comment (Erashin) [inline core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/STDCMHeuristicTests.kt:115]: ❤️

**Review by Erashin — APPROVED** [review]

File: core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/STDCMHeuristicTests.kt (hunk 3 above)
Comment (eckter) [reply core/src/test/kotlin/fr/sncf/osrd/stdcm/preprocessing/STDCMHeuristicTests.kt:84]: I amended the first commit to add a basic test
